File: src/main/dart/g0001_0100/s0079_word_search/Solution.dart
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Matrix #Backtracking
// #Algorithm_II_Day_11_Recursion_Backtracking #Big_O_Time_O(4^(m*n))_Space_O(m*n)
// #2024_10_06_Time_523_ms_(88.46%)_Space_147.3_MB_(92.31%)

class Solution {
  bool backtrace(List<List<String>> board, List<List<bool>> visited,
      String word, int index, int x, int y) {
    if (index == word.length) {
      return true;
    }
    if (x < 0 || x >= board.length || y < 0 || y >= board[0].length ||
        visited[x][y]) {
      return false;
    }
    visited[x][y] = true;

    if (word[index] == board[x][y]) {
      bool res =
          backtrace(board, visited, word, index + 1, x, y + 1) ||
              backtrace(board, visited, word, index + 1, x, y - 1) ||
              backtrace(board, visited, word, index + 1, x + 1, y) ||
              backtrace(board, visited, word, index + 1, x - 1, y);

      if (!res) {
        visited[x][y] = false;
      }
      return res;
    } else {
      visited[x][y] = false;
      return false;
    }
  }

  bool exist(List<List<String>> board, String word) {
    List<List<bool>> visited = List.generate(
        board.length, (_) => List.filled(board[0].length, false));

    for (int i = 0; i < board.length; i++) {
      for (int j = 0; j < board[0].length; j++) {
        if (backtrace(board, visited, word, 0, i, j)) {
          return true;
        }
      }
    }
    return false;
  }
}

File: src/main/swift/g0001_0100/s0079_word_search/Solution.swift
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Matrix #Backtracking
// #Algorithm_II_Day_11_Recursion_Backtracking #Top_Interview_150_Backtracking
// #Big_O_Time_O(4^(m*n))_Space_O(m*n) #2024_06_24_Time_162_ms_(96.28%)_Space_16.3_MB_(46.10%)

class Solution {
    func exist(_ board: [[Character]], _ word: String) -> Bool {
        
        var word = Array(word)

        var beginnings = [(Int, Int)]()

        for r in 0..<board.count{
            for c in 0..<board[0].count{
                if board[r][c] == word[0]{
                    beginnings.append((r, c))
                }
            }
        }

        var has = false
        for tup in beginnings{
            if word.count == 1{
                has = true
                break
            }

            var visitedBoard = [[Bool]](repeating:[Bool](repeating:false, count:board[0].count), count:board.count)
            visitedBoard[tup.0][tup.1] = true

            if proceed(tup, board, &visitedBoard, 1, word){
                has = true
                break
            }

        }

        return has

    }


    func proceed(_ startP:(Int, Int), _ board: [[Character]], _ visitedBoard:inout[[Bool]], _ targetIndex:Int, _ word:[Character]) -> Bool{

        let (r, c) = startP
        let rows = visitedBoard.count
        let cols = visitedBoard[0].count
        //try four directions
        //up
        var has = false
        if r - 1 >= 0{
            if visitedBoard[r-1][c] == false{
                if board[r-1][c] == word[targetIndex]{

                    if ((targetIndex + 1) == word.count){
                        return true
                    }

                    visitedBoard[r-1][c] = true
                    let tup = (r-1, c)
                    let res = proceed(tup, board, &visitedBoard, targetIndex+1, word)
                    visitedBoard[r-1][c] = false
                    if res == true{
                        return true
                    }
                }
            }
        }

        //left
        if c - 1 >= 0{
            if visitedBoard[r][c-1] == false{
                if board[r][c-1] == word[targetIndex]{

                    if ((targetIndex + 1) == word.count){
                        return true
                    }

                    visitedBoard[r][c-1] = true
                    let tup = (r, c-1)
                    let res = proceed(tup, board, &visitedBoard, targetIndex+1, word)
                    visitedBoard[r][c-1] = false
                    if res == true{
                        return true
                    }
                }
            }
        }

        //down
        if r + 1 < rows{
            if visitedBoard[r+1][c] == false{
                if board[r+1][c] == word[targetIndex]{

                    if ((targetIndex + 1) == word.count){
                        return true
                    }

                    visitedBoard[r+1][c] = true
                    let tup = (r+1, c)
                    let res = proceed(tup, board, &visitedBoard, targetIndex+1, word)
                    visitedBoard[r+1][c] = false
                    if res == true{
                        return true
                    }
                }
            }
        }

        //right
        if c + 1 < cols{
            if visitedBoard[r][c+1] == false{
                if board[r][c+1] == word[targetIndex]{

                    if ((targetIndex + 1) == word.count){
                        return true
                    }

                    visitedBoard[r][c+1] = true
                    let tup = (r, c+1)
                    let res = proceed(tup, board, &visitedBoard, targetIndex+1, word)
                    visitedBoard[r][c+1] = false
                    if res == true{
                        return true
                    }
                }
            }
        }

        return false
    }
}
